### PR TITLE
[PyTorch][Training][SM] PT2.1 SM Currency (cpu/cu121)

### PR DESCRIPTION
**Pull request by sirutBuasai**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
- Add sagemaker layer for cu121

### Tests run
Commit [612bb44](https://github.com/aws/deep-learning-containers/pull/3432/commits/612bb44c881f1eddfe96f3ba86693ad341e42933)
- cpu/cu121 Build
- cpu/cu121 Sanity Test
- cpu/cu121 EKS Test
- cpu/cu121 ECS Test
- cpu/cu121 SM Local Test

Commit [712926a](https://github.com/aws/deep-learning-containers/pull/3432/commits/712926a29f799d5a56a38bc74034fe53baa5055f)
- cpu/cu121 EC2 Test
- cpu/cu121 SM Remote Test (standard)

Commit [117d7a9](https://github.com/aws/deep-learning-containers/pull/3432/commits/117d7a9698a9072972842ffcd2d0b2f4fdbe66c0)
- cpu/cu121 SM Remote Test (efa)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu [hunk 1]
```diff
@@ -140,6 +140,8 @@ RUN /opt/conda/bin/mamba install -y -c conda-forge \
     pillow \
     "mpi4py>=3.1.4,<3.2" \
     h5py \
+  # dgl is not available in cuda 12+ yet
+  # && /opt/conda/bin/mamba install -y -c dglteam dgl \
```
Comment (jeet4320) [inline pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu:144]: if dglteam is not supported by cuda12+, any reason to add it here?

Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu:144]: I can just remove it tbh. We can add it later once they support it.

**Comment by jeet4320** [comment]
Please run these PR SM pytorch ddp tests on P5

File: pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu [hunk 2]
```diff
@@ -316,3 +322,163 @@ RUN rm -rf /root/.cache | true
...
+ && cd gdrcopy \
+ && sed -ie '12s@$@ -L $(CUDA)/lib/stubs@' tests/Makefile \
+ && CUDA=/opt/conda make install \
+ && rm -rf /tmp/gdrcopy
+
+# Install Nvidia Apex (needs pytorch)
+## Pin apex commit requested by sm-model-parallel team
+RUN git clone https://github.com/NVIDIA/apex \
+  && cd apex \
+  # requested by Rubik. Will wait for resolution on the illegal mem access error
```
Comment (jeet4320) [inline pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu:419]: you can remove this line as rubik is not there

**Review by jeet4320 — DISMISSED** [review]

**Review by junpuf — DISMISSED** [review]

File: pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu (hunk 2 above)
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu]: fixed

Comment (roywei) [reply pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu]: @jeet4320 @sirutBuasai we still need Apex in SM DLC for almost all PyTorch workload, for it's mix precision support and many custom fused kernels. we can install latest version and remove the version pin.

**Review by jeet4320 — APPROVED** [review]

**Pull request merged**